### ISSUE TITLE
NullInjectorError: StaticInjectorError(AppModule) for lazy loaded module with providers

### ISSUE THREAD

**Issue by MRDNZ** [angular/angular#35842]
# 🐞 bug report

### Affected Package

 The issue is probably caused by package @angular/core, but regards the lazy loading feature for Module. So it could be a compiler thing with loadChildren? 

### Is this a regression?

 Cant really tell from own experience but according to:
https://github.com/angular/angular/issues/35402
https://github.com/angular/angular/issues/35576 
which both seem to be related, i would probably say yes.

### Description

 To give some more background information:
We are using feature modules as packages and bundling these into an application to expose them as angular elements. When using these angular elements (web-components) in any other application everything works fine. So now we started migrating our application towards a new workspace where we could integrate these feature libraries and while doing so we are lazy loading the modules to improve performance.

So basically its not possible to lazy load any libraries/packages from your node_modules, therefor you have to create a small wrapper module. So we did that and use a component to render our feature. The strange thing is, if we render our angular element (web-component) everything works perfect, angular doesnt try to do smart things and the providers are available including our config. But as soon as we start to render the angular component variant, it throws error about providers that are missing.

We define our providers inside of our feature module with the providers array syntax. We are aware of the `@Injectable({ providedIn })` options, but none of these work for us.

## 🔬 Minimal Reproduction

 Because of the size i can not include any example of the current setup but i could try to make a small repro ASAP. But to give at least a bit of code/error context here is some pseudo code:

```ts
// The module from our library/package
@NgModule({
  declarations: [FeatureComponent],
  entryComponents: [FeatureComponent],
  exports: [FeatureComponent],
  providers: [FeatureService],
})
export class FeatureModule {}

// The component from our library/package
@Component({
  selector: 'feature-component',
  template: `<p>Hello World!</p>`,
})
export class FeatureComponent {
  constructor(private featureService: FeatureService) {}
}

// The lazy loaded module in the app
@NgModule({
  declarations: [LazyFeatureComponent],
  imports: [FeatureModule],
})
export class LazyFeatureModule {}

// The lazy loaded component in the app
@Component({
  selector: 'lazy-feature-component',
  template: `<feature-component></feature-component>`,
})
export class LazyFeatureComponent {}
```

## 🔥 Exception or Error

We use loadChildren to target the wrapper module from the app so it loads the feature module from the package. But when we navigate to this route we get the following error:

```
ERROR NullInjectorError: StaticInjectorError(AppModule)[FeatureComponent -> FeatureService]: 
  StaticInjectorError(Platform: core)[FeatureComponent-> FeatureService]: 
    NullInjectorError: No provider for FeatureService!
```

Which is very strange because we define our providers in the FeatureModule that is imported in the LazyFeatureModule. So it always should have the correct providers when importing it. Also the error refers to AppModule, so the FeatureComponent is not part of the FeatureModule but the AppModule/ which it shouldn't be part of.

### Workaround
If we use angular elements and put the web-component tag inside of the LazyFeatureComponent everything works fine :) Somehow the module gets properly isolated and the providers are included as expected. 

## 🌍  Your Environment

**Angular Version:**

<pre><code>
Angular CLI: 9.0.4
Node: 13.8.0
OS: win32 x64

Angular: 9.0.4
... animations, cli, common, compiler, compiler-cli, core
... elements, forms, language-service, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.4
@angular-devkit/build-angular      0.900.4
@angular-devkit/build-ng-packagr   0.900.4
@angular-devkit/build-optimizer    0.900.4
@angular-devkit/build-webpack      0.900.4
@angular-devkit/core               9.0.4
@angular-devkit/schematics         9.0.4
@angular-devkit/schematics-cli     0.900.4
@angular/cdk                       9.0.0
@angular/material                  9.0.0
@angular/material-moment-adapter   9.0.0
@ngtools/webpack                   9.0.4
@schematics/angular                9.0.4
@schematics/schematics             0.900.4
@schematics/update                 0.900.4
ng-packagr                         9.0.2
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.41.2
</code></pre>

**Anything else relevant?**

**Comment by spock123** [comment]
Isn't the entryComponents property deprecated?

**Comment by alxhub** [comment]
@MRDNZ, thank you for the report. However, I think we're going to need a reproduction in order to be able to investigate this issue.

**Comment by MRDNZ** [comment]
@spock123, I don't think it matters in our case because we don't use ivy yet for compiling due to the support for libraries. We either want ivy all the way or not at all because of the changes we have to make to be ivy compatible and we can't have both. But that's something else :)

Deprecated entryComponents indeed starts from v9 and it will probably be removed in v11.
```
Previously, the entryComponents array in the NgModule definition was used to tell the compiler which components would be created and inserted dynamically. With Ivy, this isn't a requirement anymore and the entryComponents array can be removed from existing module declarations.
````

But for now for our use case this won't be a problem yet.

**Comment by MRDNZ** [comment]
@alxhub, thanks! I will double check our issue with this information, it does seem to be a similar problem.

Also i will try to make a small reproduction this week.

**Comment by mlc-mlapis** [comment]
Probably related to https://github.com/angular/angular/issues/35576?

**Comment by MRDNZ** [comment]
So while creating a simple reproduction, i could not get the same error that we got in the full setup. But i did find the cause of our problem, somehow the following combination of a Route config: 
```
{
  path: 'some-path',
  loadChildren: () => import('../some/path.module').then(m => m.SomePathModule),
  component: SomePathComponent,
}
``` 

Where there are both a loadChildren and component are defined, didnt show up as an error inside of the setup. In the simple reproduction i did got an error. After resolving the error everything worked just fine. Also it does make sense but i guess we overlooked this due to a different error message and the quick workaround with using a web-component.

Thanks anyways! 😃
